Add stylelint for SCSS linting

noisysocks opened the pull request:
Supersedes https://github.com/WordPress/gutenberg/pull/1151. Closes #8618.

Adds the `npm run lint-css` and `npm run lint-css:fix` commands for SCSS linting, and configures `npm run lint` to automatically run `npm run lint-css`.

Linting is performed by [stylelint](https://github.com/stylelint/stylelint). I've configured stylelint to extend [stylelint-config-wordpress](https://github.com/WordPress-Coding-Standards/stylelint-config-wordpress) which enforces the [WordPress CSS coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/css/).

To save time and reduce merge conflicts, I've disabled every rule that isn't totally trivial to fix (e.g. with a find and replace). We can work towards re-enabling these rules one by one over the medium term.

To test: 

1. `npm run lint-css`
2. Open the demo post and check for visual regressions
3. Check the diff and make sure `stylelint --fix` or one of my regular expressions didn't completely backfire

noisysocks commented on core-blocks/button/editor.scss:
```diff
@@ -1,9 +1,11 @@
 .editor-block-list__block[data-type="core/button"] {
+
```
Weird... why does it want spaces here, @ntwb?

ntwb replied:
A few reasons:

• Due to the number of permutations and combinations of _where_ and _where not_ new empty lines should occur, I've an issue to audit all of these at https://github.com/WordPress-Coding-Standards/stylelint-config-wordpress/issues/65

• The lack off SCSS used currently in WordPress leading to not having an official "WordPress SCSS Coding Standards"

-----

Solution/Workaround:

Update the config used here to override the appropriate rules here in this PR and then follow up by updating `stylelint-config-wordpress`, either or both the CSS and SCSS configs and releasing a new version of `stylelint-config-wordpress`.

ntwb commented on packages/components/src/button/style.scss:
```diff
@@ -60,56 +61,62 @@
+				color(theme(primary) shade(30%)) 72%,
+				theme(primary) 72%
+			) !important;
+			/* stylelint-enable */
```
Note to self: Investigate supporting this in stylelint core

ntwb commented on edit-post/components/meta-boxes/meta-boxes-area/style.scss:
```diff
@@ -28,7 +29,9 @@
-	#poststuff h2.hndle { /* WordPress selectors yolo */
+	#poststuff h2.hndle {
+
+		/* WordPress selectors yolo */
```
This is not ideal, it's an idiosyncratic stylelint issue that requires fixing 😞

ntwb reviewed (approved):
Awesome work @noisysocks 💯

ntwb commented:
Reviewing the diff [without whitespace changes](https://github.com/WordPress/gutenberg/pull/8647/files?w=1) is helpful 👌

gziolo commented:
Any idea why one of the unit tests failed on Travis?

tofumatt reviewed (approved):
All of those weird new empty newlines aren't ideal, but I prefer automated, written rules I don't like over unwritten, unautomated ones I do like.

I'd be cool to fix all those newlines (eg https://github.com/WordPress/gutenberg/pull/8647/files#diff-44e5de5ab7a2465d8a970cbd777c427eR2). But hey, rules are cool!

noisysocks commented:
Thanks all! I disabled `comment-empty-line-before`, `rule-empty-line-before`, and `at-rule-empty-line-before` as quick fix for the weird whitespace that was being added. Better to have the bare minimum than nothing at all 🙂

noisysocks commented:
<img width="284" alt="screen shot 2018-08-08 at 15 01 15" src="https://user-images.githubusercontent.com/612155/43817275-ebe03f62-9b1b-11e8-831c-d8970564badf.png">

That ain't right... 😅

noisysocks committed:
Add stylelint

Adds stylelint and configures it to lint SCSS files with a very
conservative rule set.

noisysocks committed:
Make "npm run lint" run its tasks concurrently

gziolo commented on package.json:
```diff
@@ -161,11 +163,13 @@
 		"fixtures:generate": "npm run fixtures:server-registered && cross-env GENERATE_MISSING_FIXTURES=y npm run test-unit",
 		"fixtures:regenerate": "npm run fixtures:clean && npm run fixtures:generate",
-		"lint": "npm run lint-js && npm run lint-pkg-json",
+		"lint": "npm run lint-js && npm run lint-pkg-json && npm run lint-css",
```
Can we run it concurrently?

noisysocks replied:
👍 b97714574

gziolo reviewed (approved):
Awesome one 👏 

I ❤️ `npm run lint-css:fix` the most!

ntwb commented:
❤️

aduth commented:
In case anyone is like me and was looking for editor integration for Sublime Text, here's the plugin for use with [SublimeLinter](https://github.com/SublimeLinter/SublimeLinter):

https://github.com/SublimeLinter/SublimeLinter-stylelint

youknowriad commented:
Who uses sublime nowadays :trollface: ?

tofumatt commented:
It's way faster than Atom! :trollface: :trollface: :trollface:

youknowriad commented:
@tofumatt I fully deserve this one :P
